pkg/github/scim.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2025 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/v67/github"
+)
+
+// SCIMClient handles direct HTTP communication with the GHES SCIM API.
+// API doc: https://docs.github.com/en/enterprise-server@3.17/admin/managing-iam/provisioning-user-accounts-with-scim/provisioning-users-and-groups-with-scim-using-the-rest-api#provisioning-users-with-the-rest-api
+type SCIMClient struct {
+	httpClient *http.Client
+	baseURL    *url.URL
+}
+
+// NewSCIMClient creates a new client for the GHES SCIM API.
+func NewSCIMClient(httpClient *http.Client, baseURL string) (*SCIMClient, error) {
+	u, err := url.Parse(strings.TrimSuffix(baseURL, "/") + "/api/v3/scim/v2/")
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse base url %q: %w", baseURL, err)
+	}
+	return &SCIMClient{httpClient: httpClient, baseURL: u}, nil
+}
+
+// ListUsers fetches all SCIM provisioned users from the enterprise, handling SCIM pagination.
+func (c *SCIMClient) ListUsers(ctx context.Context) (map[string]*github.SCIMUserAttributes, error) {
+	allUsers := make(map[string]*github.SCIMUserAttributes)
+	startIndex := 1
+	for {
+		path := "Users"
+		rel, err := url.Parse(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse path %q: %w", path, err)
+		}
+		q := rel.Query()
+		q.Set("startIndex", strconv.Itoa(startIndex))
+		q.Set("count", "100")
+		rel.RawQuery = q.Encode()
+
+		var result github.SCIMProvisionedIdentities
+		if _, err := c.do(ctx, http.MethodGet, c.baseURL.ResolveReference(rel).String(), nil, &result); err != nil {
+			return nil, fmt.Errorf("failed to list scim users starting at index %d: %w", startIndex, err)
+		}
+
+		if len(result.Resources) == 0 {
+			break
+		}
+		for _, u := range result.Resources {
+			allUsers[u.UserName] = u
+		}
+		if len(allUsers) >= *result.TotalResults {
+			break
+		}
+		startIndex += len(result.Resources)
+	}
+	return allUsers, nil
+}
+
+// CreateUser provisions a new user.
+func (c *SCIMClient) CreateUser(ctx context.Context, user *github.SCIMUserAttributes) (*github.SCIMUserAttributes, *github.Response, error) {
+	path := "Users"
+	var createdUser github.SCIMUserAttributes
+	resp, err := c.do(ctx, http.MethodPost, c.baseURL.ResolveReference(&url.URL{Path: path}).String(), user, &createdUser)
```

Comment:
I think I'm seeing two different ways of constructing the Users path. In ListUsers, call `url.Parse("Users")`. Here, we use `&url.URL{Path: "Users"}` . I think these should be the same because they're doing the same thing. I'm fine with either approach.